runtimes/go-1.22/src/main.go

```diff
@@ -0,0 +1,307 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"openruntimes/handler"
+	"openruntimes/types"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func action(w http.ResponseWriter, r *http.Request) error {
+	timeout := r.Header.Get("x-open-runtimes-timeout")
+
+	var safeTimeout int
+
+	if timeout != "" {
+		safeTimeoutInteger, err := strconv.Atoi(timeout)
+
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			w.Header().Set("content-type", "text/plain")
+			fmt.Fprintf(w, "Header \"x-open-runtimes-timeout\" must be an integer greater than 0.")
+			return nil
+		}
+
+		safeTimeout = safeTimeoutInteger
+	}
+
+	secret := r.Header.Get("x-open-runtimes-secret")
+	serverSecret := os.Getenv("OPEN_RUNTIMES_SECRET")
+
+	if secret == "" || secret != serverSecret {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Header().Set("content-type", "text/plain")
+		fmt.Fprintf(w, "Unauthorized. Provide correct \"x-open-runtimes-secret\" header.")
+		return nil
+	}
+
+	contentType := r.Header.Get("content-type")
+	if contentType == "" {
+		contentType = "text/plain"
+	}
+
+	bodyBytes, err := io.ReadAll(r.Body)
+
+	if err != nil {
+		return errors.New("Could not parse body into a string.")
+	}
+
+	bodyRaw := string(bodyBytes)
+
+	var body interface{} = bodyRaw
+
+	if contentType == "application/json" {
+		if bodyRaw != "" {
+			err := json.Unmarshal([]byte(bodyRaw), &body)
+
+			if err != nil {
+				return errors.New("Could not parse body into a JSON.")
+			}
+		} else {
+			body = map[string]interface{}{}
+		}
+	}
+
+	headers := map[string]string{}
+	for key, value := range r.Header {
+		key = strings.ToLower(key)
+
+		if strings.HasPrefix(key, "x-open-runtimes-") == false {
+			headers[key] = value[0]
+		}
+	}
+
+	method := r.Method
+
+	scheme := "http"
+	schemeHeader := r.Header.Get("x-forwarded-proto")
+	if schemeHeader != "" {
+		scheme = schemeHeader
+	}
+
+	defaultPort := 80
+	if scheme == "https" {
+		defaultPort = 443
+	}
+
+	var host string
+	var port int
+
+	hostHeader := r.Host
+
+	if strings.Contains(hostHeader, ":") {
+		chunks := strings.SplitN(hostHeader, ":", 2)
+		host = chunks[0]
+		parsedPort, err := strconv.Atoi(chunks[1])
+
+		if err != nil {
+			return errors.New("Host header could not be parsed into host and port.")
+		}
+
+		port = parsedPort
+	} else {
+		host = hostHeader
+		port = defaultPort
+	}
+
+	path := r.URL.Path
+
+	queryString := r.URL.RawQuery
+
+	query := map[string]string{}
+	queryChunks := strings.Split(queryString, "&")
+	for _, chunk := range queryChunks {
+		queryChunk := strings.SplitN(chunk, "=", 2)
+
+		queryKey := ""
+		queryValue := ""
+
+		if len(queryChunk) > 0 {
+			queryKey = queryChunk[0]
+		}
+
+		if len(queryChunk) > 1 {
+			queryValue = queryChunk[1]
+		}
+
+		if queryKey != "" {
+			query[queryKey] = queryValue
+		}
+	}
```

Comment:
you can use https://pkg.go.dev/net/url#ParseQuery 👍🏻

Reply:
Applied